src/model_config_utils.cc

```diff
@@ -727,66 +708,136 @@ NormalizeModelConfig(
     if (!optimization->has_output_pinned_memory()) {
       optimization->mutable_output_pinned_memory()->set_enable(true);
     }
+  }
 
-    // Make sure there is at least one instance_group.
-    if (config->instance_group().size() == 0) {
-      inference::ModelInstanceGroup* group = config->add_instance_group();
-      group->set_name(config->name());
-    }
+  return Status::Success;
+}
 
-    // Creates a set of supported GPU device ids
-    std::set<int> supported_gpus;
+Status
+NormalizeInstanceGroup(
+    const double min_compute_capability,
+    const std::vector<inference::ModelInstanceGroup>& preferred_groups,
+    inference::ModelConfig* config)
+{
+  // Instance group setting doesn't apply to ensemble
+  if (config->has_ensemble_scheduling()) {
+    return Status::Success;
+  }
+
+  // Creates a set of supported GPU device ids
+  std::set<int> supported_gpus;
 #ifdef TRITON_ENABLE_GPU
-    // Get the total number of GPUs from the runtime library.
-    Status status = GetSupportedGPUs(&supported_gpus, min_compute_capability);
-    if (!status.IsOk()) {
-      return status;
-    }
+  // Get the total number of GPUs from the runtime library.
+  Status status = GetSupportedGPUs(&supported_gpus, min_compute_capability);
+  if (!status.IsOk()) {
+    return status;
+  }
 
 #endif  // TRITON_ENABLE_GPU
 
-    // Assign default name, kind and count to each instance group that
-    // doesn't give those values explicitly. For KIND_GPU, set GPUs to
-    // all available if not specified explicitly.
-    size_t cnt = 0;
-    for (auto& group : *config->mutable_instance_group()) {
-      // Name
-      if (group.name().empty()) {
-        group.set_name(config->name() + "_" + std::to_string(cnt));
-      }
-      cnt++;
-
-      // For KIND_AUTO... if there are no GPUs or if any of the listed
-      // 'gpu's are not present, then use KIND_CPU.
-      if (group.kind() == inference::ModelInstanceGroup::KIND_AUTO) {
+  // Make sure there is at least one instance_group.
+  if (config->instance_group().empty()) {
+    inference::ModelInstanceGroup* group = config->add_instance_group();
+    group->set_name(config->name());
+
+    for (const auto& pg : preferred_groups) {
+      group->set_kind(pg.kind());
```

Comment:
Do we need to have a multi-GPU test for this to make sure that the instances get correctly populated?

Reply:
Can you elaborate?

Reply:
I meant to add a multi-GPU test that tries the different combinations of the preferred instance group and model instance group if you think it make sense.